datafusion/common/src/tree_node.rs

```diff
@@ -289,15 +492,80 @@ pub enum RewriteRecursion {
     Skip,
 }
 
-/// Controls how the [`TreeNode`] recursion should proceed for [`TreeNode::visit`].
+/// Controls how the [`TreeNode`] recursion should proceed for [`TreeNode::visit_down()`] and
+/// [`TreeNode::visit()`].
 #[derive(Debug)]
-pub enum VisitRecursion {
-    /// Continue the visit to this node tree.
+pub enum TreeNodeRecursion {
+    /// Continue the visit to the next node.
     Continue,
-    /// Keep recursive but skip applying op on the children
-    Skip,
-    /// Stop the visit to this node tree.
+
+    /// Prune the current subtree.
+    /// If a preorder visit of a tree node returns [`TreeNodeRecursion::Prune`] then inner
+    /// children and children will not be visited and postorder visit of the node will not
+    /// be invoked.
+    Prune,
+
+    /// Stop recursion on current tree.
+    /// If recursion runs on an inner tree then returning [`TreeNodeRecursion::Stop`] doesn't
+    /// stop recursion on the outer tree.
     Stop,
+
+    /// Stop recursion on all (including outer) trees.
+    StopAll,
+}
+
+impl TreeNodeRecursion {
+    pub fn and_then_on_continue<F>(self, f: F) -> Result<TreeNodeRecursion>
+    where
+        F: FnOnce() -> Result<TreeNodeRecursion>,
+    {
+        match self {
+            TreeNodeRecursion::Continue => f(),
+            o => Ok(o),
+        }
+    }
+
+    pub fn continue_on_prune(self) -> Result<TreeNodeRecursion> {
+        Ok(match self {
+            TreeNodeRecursion::Prune => TreeNodeRecursion::Continue,
+            o => o,
+        })
+    }
+
+    pub fn fail_on_prune(self) -> Result<TreeNodeRecursion> {
```

Comment:
I don't understand the usecase for this method -- if there is going to be a panic, perhaps it would be clearer to put the check directly at the callsite with a explination for why the situation warrants a panic